src/applications/bmqstoragetool/m_bmqstoragetool_parameters.h

```diff
@@ -0,0 +1,272 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// m_bmqtool_parameters.h                                             -*-C++-*-
+#ifndef INCLUDED_M_BMQTOOL_PARAMETERS
+#define INCLUDED_M_BMQTOOL_PARAMETERS
+
+//@PURPOSE: Provide a class holding command line parameters for 'bmqtool'.
+//
+//@CLASSES:
+//  m_bmqtool::ParametersVerbosity: enum for verbosity mode.
+//  m_bmqtool::ParametersMode     : enum for the tool mode.
+//  m_bmqtool::ParametersLatency  : enum for latency mode.
+//  m_bmqtool::Parameters         : holds all parameter values
+//
+//@DESCRIPTION: This component provides a value-semantic type holding the
+// command-line parameters for the 'bmqtool' program.
+
+// bmqstoragetool
+#include <m_bmqstoragetool_queuemap.h>
+
+// MQB
+#include <mqbs_datafileiterator.h>
+#include <mqbs_filestoreprotocol.h>
+#include <mqbs_filesystemutil.h>
+#include <mqbs_journalfileiterator.h>
+#include <mqbs_mappedfiledescriptor.h>
+
+// MWC
+#include <mwcu_stringutil.h>
+
+// BDE
+#include <bsl_iosfwd.h>
+#include <bsl_string.h>
+#include <bslma_allocator.h>
+#include <bsls_types.h>
+
+namespace BloombergLP {
+
+namespace m_bmqstoragetool {
+
+// ================
+// class Parameters
+// ================
+
+struct CommandLineArguments {
+    // PUBLIC DATA
+    bsls::Types::Int64 d_timestampGt;
+    // Filter messages by minimum timestamp
+    bsls::Types::Int64 d_timestampLt;
+    // Filter messages by maximum timestamp
+    bsl::string d_journalPath;
+    // Path to find all files from
+    bsl::string d_journalFile;
+    // Path to read journal files from
+    bsl::string d_dataFile;
+    // Path to read data files from
+    bsl::string d_cslFile;
+    // Path to read CSL files from
+    bsl::vector<bsl::string> d_guid;
+    // Filter messages by message guids
+    bsl::vector<bsl::string> d_queueKey;
+    // Filter messages by queue keys
+    bsl::vector<bsl::string> d_queueName;
+    // Filter messages by queue names
+    int d_dumpLimit;
+    // Limit number of bytes to
+    bool d_details;
+    // Print message details
+    bool d_dumpPayload;
+    // Print message payload
+    bool d_summary;
+    // Print summary of messages
+    bool d_outstanding;
+    // Show only outstanding messages (not deleted)
+    bool d_confirmed;
+    // Show only messages, confirmed by all the appId's
+    bool d_partiallyConfirmed;
+    // Show only messages, confirmed by some of the appId's
+
+    // CREATORS
+    explicit CommandLineArguments(bslma::Allocator* allocator = 0);
+
+    // MANIPULATORS
+    /// Validate the consistency of all settings.
+    bool validate(bsl::string* error);
+};
+
+class Parameters {
+  public:
+    // MANIPULATORS
+    virtual mqbs::JournalFileIterator* journalFileIterator() = 0;
+    virtual mqbs::DataFileIterator*    dataFileIterator()    = 0;
+
+    // ACCESSORS
+    virtual bsls::Types::Int64       timestampGt() const        = 0;
```

Comment:
Having _virtual_ accessors to something which looks like a simple set of values (moreover when it is named "Parameters") looks a bit suspicious. The key point is that a program should consist of a set of easily understandable abstractions. So you usually implement different actors with their own implementations of virtual methods of some interface, and each of them has their own parameters that are represented by some struct-like type. The struct-like type is something like this:
```
struct Entity1Parameters {
  Foo f;
  Bar b;
};
```
or
```
class Entity1Parameters {
public:
  Foo GetF();
  void SetF(const Foo& f);
  Bar GetB();
  void SetB(const Bar& f);
  
};
```

Reply:
But the `ParametersMock` consists of ".WillRepeatedly(Return(foo))". You could just create a struct, and set some values required for testing to each field.

Reply:
We can't get rid of polymorphism at all here because file-related members (FileHandler<...>) in ParametersReal class deal with real files. Of course, I can move other members with their getters to the base class, but it will have some undesirable consequences:
1) base class Parameters will no longer be interface class
2) all its members will be initialised by default, including those ones, that are not used in a particular test

Reply:
No, please don't inherit from "Parameters"! 
You need some external entity though: mixing together something that has a semantic of "config parameter" with something that has a state, this state is updated while reading etc is pretty bad idea. If a file is a part of the config, the only option to refer to in a struct _describing parameters_ is just its filename.

Reply:
Ok, it looks sensible. What do you think about spliting **Parameters** class into two parts? 
1. an abstract class... lets say **FileManager** responsible for file-specific logic with two realisations (one for real files and another for tests)
2. POD struct for the rest of parameters
We can pass both of them to **CommandProcessor**'s contructor then

Reply:
this sounds reasonable, go for it!